Fractionated Morse cipher

dex-r opened the pull request:
This PR implements the fractionated morse cipher. I said in Issue #14 that it would require alphanumeric alphabets, but its turns out Morse code supports a very specific set of special characters as well, so I've kept the alphabet used inside `fractionatedmorse.rs`.

arosspope commented on src/common/keygen.rs:
```diff
@@ -0,0 +1,84 @@
+
+    // Loop over the key and add each unique letter to the keyed alphabet.
+    for c in key.chars() {
+        match find_position(c) {
```
To make it uniform across the codebase (and more easy to read) could you change this to `alphabet::find_position(c)`, same with `get_letter`

arosspope commented on src/common/keygen.rs:
```diff
@@ -0,0 +1,84 @@
+/// e.g. A key of `alphabet` will produce the result `alphbetcdfgijkmnoqrsuvwxyz`.
+///
+/// Will return Err if a non-alphabetic symbol is present in the key.
+pub fn generate_keyed_alphabet(key: &str, make_uppercase: bool) -> Result<String, &'static str> {
```
perhaps rename this to `keyed_alphabet` and callers would invoke it as `keygen::keyed_alphabet`

arosspope commented on src/fractionatedmorse.rs:
```diff
@@ -0,0 +1,333 @@
+use std::ascii::AsciiExt;
+
+// The Morse alphabet.
+const MORSE_ALPHABET: [(char, &str); 50] = [
```
What resource did you use to generate this alphabet? (Perhaps include a link in the comment).

arosspope replied:
Thinking about it, I would prefer if this was moved to its own file in common called `morse_alphabet` with similar access functions to the `alphabet` module.

arosspope commented on src/fractionatedmorse.rs:
```diff
@@ -0,0 +1,333 @@
+    fn encrypt_morse(message: String) -> Result<String, &'static str> {
+        let mut morse = String::new();
+
+        // Convert each letter in message to corresponding Morse characters.
```
Correct me if I am wrong, but whitespace is not a valid character in the morse alphabet. Because of this, I believe that we should reject messages that contain whitespace. The caller would have to make the decision of whether to strip the whitespace themselves or use a different cipher. I made a similar decision in the hill cipher.

dex-r replied:
You're right that you can't really encode a space, but I guess it's implied by the double seperator. I'm really basing this all off what the [dcode page](http://www.dcode.fr/fractionated-morse) does. Also, if you think of the seperator as a pause in the sending of a morse message, a short pause `|` would represent the end of a character and a double pause, a long pause, `||` would be the end of a word. I do think it makes sense, but i'm happy to remove it if you wish.

arosspope replied:
Ideally I want to avoid the scenario `m !== decrypt(encrypt(m))` where m is a message with spaces in it. 

Thats why I think its better to reject.. However if you find someway to encode whitespace in morsecode to preserve it, then perhaps we could keep it?

dex-r replied:
Yeah, I can't consistantly encode all whitespace, so requiring it to be removed might be best.

arosspope commented on src/fractionatedmorse.rs:
```diff
@@ -0,0 +1,333 @@
+    /// separator. The Morse code is ended with two separators `||`. This function returns `Err`
+    /// if an unsupported symbol is present. The support characters are `a-z`, `A-Z`, `0-9`, spaces
+    /// and the special characters `@ ( ) . , : ' " ! ? - ; =`.
+    fn encrypt_morse(message: String) -> Result<String, &'static str> {
```
Could you call this function `to_morse` instead as you're not really encrypting here, just converting to morse code.

arosspope replied:
Actually the fractionation of the morse occurs here right? So maybe this should be called `to_fractionated_morse`. If this is the case, I think you should do the padding here as well.

dex-r replied:
I think `to_morse` would be best. The thing is the morse string and the fractionated morse string are actually the same, you just interpret them differently. To get the morse you split the message by the seperator, `|`, and to get fractionated morse you split the message into trigraphs.

dex-r replied:
Now that I think about it, the morse strings and fractionated morse strings are different but only because of the padding, so if the padding was added in this function then the name `to_fractionated_morse` would make more sense, whereas if the padding is added where it currently is, the name `to_morse` would be better for this function.

arosspope replied:
Either works for me. Whatever you think makes more sense

arosspope commented on src/fractionatedmorse.rs:
```diff
@@ -0,0 +1,333 @@
+        // in-between each word.
+        for word in message.split_whitespace() {
+            for c in word.chars() {
+                if let Some(pos) = MORSE_ALPHABET.iter().position(|m| m.0 == c.to_ascii_lowercase()) {
```
Why bother importing the 'ascii' library? I believe the function `to_lowercase()` in the standard char library (no import required) should suffice

dex-r replied:
It is required. You get an error:
 `help: items from traits can only be used if the trait is in scope; the following trait is implemented but not in scope, perhaps add a 'use' for it:`
Don't really understand that all too well, but i do what the compiler tells me ;)

dex-r replied:
oh i see what you mean

dex-r replied:
I remember now, I did look at that function but it returns an iterator not a char, and i couldn't see a consise way to convert it directly into a char for comparison. Do you know how that might be done?

arosspope replied:
Perhaps you could add a 3rd element to the tuple in the morse code alphabet - the upper case version of the character (kinda like how I have the uppercase version of the normal alphabet). 

For symbols and numbers you could just repeat it where the uppercase version of it would go.

arosspope replied:
Actually, rather than using an array of tuples to store the morse code alphabet, have you considered using a hashmap?

dex-r replied:
I did consider a hashmap, but it requires an external crate to get a macro for creating one (the equivilant of vec!). Seemed a bit extreme, but it would be handy.

arosspope replied:
All good, leave it for now and I can do some investigating

arosspope commented on src/fractionatedmorse.rs:
```diff
@@ -0,0 +1,333 @@
+    /// Takes a Morse string, pads it with dots to a length that is a multiple of 3, and converts
+    /// it to an alphabetic string. This function returns `Err` if an invalid Fractionated Morse
+    /// trigraph is encountered.
+    fn decrypt_frac_morse(keyed_alphabet: &String, mut message: String) -> Result<String, &'static str> {
```
- I think this should be called `fractionate_morse` as the current name is a bit confusing when used in the context of the encrypt function.
- change `message` to `morse`
- change `&String` to `&str`

arosspope replied:
Please ignore my first dotpoint.

arosspope commented on src/fractionatedmorse.rs:
```diff
@@ -0,0 +1,333 @@
+        let mut ciphertext = String::new();
+
+        // Trigraphs are required, so pad the message with dots until the length is a multiple of 3.
+        while message.len() % 3 != 0 {
```
I think rather than making `message` a mutable variable in this function, this padding code should be moved to its own function.

arosspope commented on src/fractionatedmorse.rs:
```diff
@@ -0,0 +1,333 @@
+
+    /// Takes a alphabetic string and converts it to Fractionated Morse. This function will
+    /// return `Err` if a non-alphabetic symbol is present in the message.
+    fn encrypt_frac_morse(keyed_alphabet: &String, message: String) -> Result<String, &'static str> {
```
perhaps change this to `to_fractionated_morse` to make it clear that we are going from text -> fractionated_morse

dex-r commented:
Ok, so I've done most of the changes. I've moved the morse alphabet to its own file, but it doesn't have access functions, only conversion functions since those were all i needed. I've also added uppercase letters as entirely new tuples in the alphabet and just repeated the Morse decodings. I agree with your comment that hashmaps would be good if you do discover a way to do them.

Also, the final `Cargo.toml` and `README` version updates are done in the merge commit above 3e82233.
